gtsfm/retriever/netvlad_retriever.py

```diff
@@ -0,0 +1,262 @@
+        block_j_query_descs = torch.from_numpy(np.array(block_j_query_descs))
+
+        # Einsum equivalent to (img_descs @ img_descs.T)
+        sim_block = torch.einsum("id,jd->ij", block_i_query_descs.to(device), block_j_query_descs.to(device))
+        # sim[i_start:i_end, j_start:j_end] = sim_block
+
+        return SubBlockSimilarityResult(i_start=i_start, i_end=i_end, j_start=j_start, j_end=j_end, subblock=sim_block)
+
+    def _aggregate_subblocks(self, subblock_results: List[SubBlockSimilarityResult], num_images: int) -> torch.Tensor:
+        """Aggregate results from many independently computed sub-blocks of the similarity matrix into a single matrix.
+
+        Args:
+            subblock_results: metadata and results of similarity matrix sub-block computation.
+            num_images: number of images to compare for matching.
+
+        Returns:
+            sim: tensor of shape (num_images, num_images) representing similarity matrix.
+        """
+        sim = torch.zeros((num_images, num_images))
+
+        for sr in subblock_results:
+            sim[sr.i_start : sr.i_end, sr.j_start : sr.j_end] = sr.subblock
+        return sim
+
+    def compute_pairs_from_similarity_matrix(
+        self, sim: torch.Tensor, loader: LoaderBase, visualize: bool = True
+    ) -> List[Tuple[int, int]]:
+        """
+        Args:
+            sim: tensor of shape (num_images, num_images) representing similarity matrix.
+            loader: image loader. The length of this loader will provide the total number of images
+                for exhaustive global descriptor matching.
+            visualize: whether to save a visual plot of the computed image similarity matrix.
+
+        Returns:
+            pair_indices: (i1,i2) image pairs.
+        """
+        num_images = len(loader)
+        query_names = loader.image_filenames()
+        # Avoid self-matching and disallow lower triangular portion
+        is_invalid_mat = ~np.triu(np.ones((num_images, num_images), dtype=bool))
+        np.fill_diagonal(a=is_invalid_mat, val=True)
+        pairs = pairs_from_score_matrix(
+            sim, invalid=is_invalid_mat, num_select=self._num_matched, min_score=self._min_score
+        )
+
+        if visualize:
+            plt.imshow(np.triu(sim.detach().cpu().numpy()))
+            os.makedirs(PLOT_SAVE_DIR, exist_ok=True)
+            plt.title("Image Similarity Matrix")
+            np.savetxt(
+                fname=os.path.join(PLOT_SAVE_DIR, "netvlad_similarity_matrix.txt"),
+                X=sim.detach().cpu().numpy(),
+                fmt="%.2f",
+                delimiter=",",
+            )
+            plt.savefig(os.path.join(PLOT_SAVE_DIR, "netvlad_similarity_matrix.jpg"), dpi=500)
+            plt.close("all")
+
+        named_pairs = [(query_names[i], query_names[j]) for i, j in pairs]
+        logger.info("Found %d pairs from the NetVLAD Retriever.", len(pairs))
+        logger.info("Image Name Pairs:" + str(named_pairs))
```

Comment:
Should probably remove this before merging

Reply:
unresolved

Reply:
This is only logged once so I'm fine with leaving it